Add RFC7523 compliant OAuth access token endpoint

chdorner opened the pull request:
This endpoint handles the exchange of a JWT authorization grant to an
API access token as described in [RFC7523](https://tools.ietf.org/html/rfc7523).

According to the OAuth 2.0 ([RFC6749](https://tools.ietf.org/html/rfc6749)) spec the token endpoint needs to handle `POST` requests. We currently already have a cookie-based token exchange implemented under `/api/token`, but this one is only called via `GET`. So I made sure that the current implementation only listens to `GET` and started the new implementation in a different views file only listening to `POST`.

Partners will generate JWT tokens for their users and sign them with
their client secret. They then pass that JWT token as a configuration to
the Hypothesis client (that part still needs work), and it will then
exchange the JWT authorization grant for an access token via the `POST
/api/token` endpoint.

RFC7523 [requires the use of the RS256 algorithm](https://hyp.is/1izOopB8EeaKcHPHrD9aYw/tools.ietf.org/html/rfc7523) to sign the JWT token,
the implementation in this commit still uses thi HS256 algorithm. In the
future, authclients would generate a public/private keypair and share
the public part with us.

chdorner commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Implement token issuing endpoint](https://trello.com/c/Qtx7dZT0/7-implement-token-issuing-endpoint)

nickstenning commented on h/views/api_auth.py:
```diff
@@ -0,0 +1,114 @@
+    return {
+        'access_token': token.value,
+        'token_type': 'bearer',
+        'expires_in': expires_in.seconds,
```
I think you probably want to use `expires_in.total_seconds()` here, because while this will currently work, it'll return zero if you were to change `expires_in` to `timedelta(days=1)`.

nickstenning commented on h/views/api_auth.py:
```diff
@@ -0,0 +1,114 @@
+        raise OAuthTokenError('invalid_grant',
+                              description='user with userid described in subject could not be found')
+
+    expires_in = timedelta(hours=1)
```
It's probably worth pulling this out into a module "constant".

nickstenning commented on h/views/api_auth.py:
```diff
@@ -0,0 +1,114 @@
+    grant_token = _validate_request(request)
+    userid = _extract_userid(request, grant_token)
+
+    user = request.db.query(models.User).filter_by(userid=userid).one_or_none()
```
It's probably a good idea to use `UserService.fetch` to retrieve users by userid.

nickstenning commented on h/views/api_auth.py:
```diff
@@ -0,0 +1,114 @@
+    request.response.status_code = context.status_code
+    resp = {'error': context.message}
+    if context.description:
+        resp['error_description'] = context.description
```
I think the name for this that we use elsewhere is `reason`.

seanh replied:
I think `'error_description'` is mandated by RFC7523 though..

chdorner replied:
Sean is right (RFC6749, OAuth, to be exact).
https://tools.ietf.org/html/rfc6749#appendix-A.8

chdorner commented:
I've addressed most comments here and shuffled a few things around. I will implement the CORS decorator tomorrow and then create at least two new PR's so we can split this up a bit.

chdorner commented:
Closing this in favour of a series of smaller pull requests.
